src/util/psbt/map/global.rs

```diff
@@ -0,0 +1,133 @@
+use util::psbt;
+use util::psbt::Error;
+
+/// A key-value map for global data.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Global {
+    /// The unsigned transaction, scriptSigs and witnesses for each input must be
+    /// empty.
+    pub unsigned_tx: Transaction,
+    /// Unknown global key-value pairs.
+    pub unknown: HashMap<raw::Key, Vec<u8>>,
+}
+
+impl Global {
+    /// Create a Global from an unsigned transaction, error if not unsigned
+    pub fn from_unsigned_tx(tx: Transaction) -> Result<Self, encode::Error> {
+        for txin in &tx.input {
+            if !txin.script_sig.is_empty() {
+                return Err(Error::UnsignedTxHasScriptSigs.into());
+            }
+
+            if !txin.witness.is_empty() {
+                return Err(Error::UnsignedTxHasScriptWitnesses.into());
+            }
+        }
+
+        Ok(Global {
+            unsigned_tx: tx,
+            unknown: Default::default(),
+        })
+    }
+}
+
+impl Map for Global {
+    fn insert_pair(&mut self, pair: raw::Pair) -> Result<(), encode::Error> {
+        let raw::Pair {
+            key: raw_key,
+            value: raw_value,
+        } = pair;
+
+        match raw_key.type_value {
+            0u8 => {
+                return Err(Error::DuplicateKey(raw_key).into());
+            }
+            _ => {
+                if self.unknown.contains_key(&raw_key) {
+                    return Err(Error::DuplicateKey(raw_key).into());
+                } else {
+                    self.unknown.insert(raw_key, raw_value);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn get_pairs(&self) -> Result<Vec<raw::Pair>, encode::Error> {
+        let mut rv: Vec<raw::Pair> = Default::default();
+
+        rv.push(raw::Pair {
+            key: raw::Key {
+                type_value: 0u8,
+                key: vec![],
+            },
+            value: self.unsigned_tx.serialize(),
+        });
+
+        for (key, value) in self.unknown.iter() {
+            rv.push(raw::Pair {
+                key: key.clone(),
+                value: value.clone(),
+            });
+        }
+
+        Ok(rv)
+    }
+
+    fn merge(&mut self, other: Self) -> Result<(), psbt::Error> {
+        if self.unsigned_tx != other.unsigned_tx {
+            return Err(psbt::Error::UnexpectedUnsignedTx {
+                expected: self.unsigned_tx.clone(),
+                actual: other.unsigned_tx,
+            });
+        }
+
+        self.unknown.extend(other.unknown);
+        Ok(())
+    }
+}
+
+impl_psbtmap_consensus_encoding!(Global);
+
+impl<D: Decoder> Decodable<D> for Global {
+    fn consensus_decode(d: &mut D) -> Result<Self, encode::Error> {
+        let mut pairs: Vec<raw::Pair> = Default::default();
+
+        while let Ok(pair) = Decodable::consensus_decode(d) {
+            pairs.push(pair);
+        }
+
+        let (tx_raw_kv_pair, unknown_raw_kv_pairs): (
+            Vec<raw::Pair>,
+            Vec<raw::Pair>,
+        ) = pairs
```

Comment:
In general I'm not thrilled with this design where we allocate a vector of pairs, allocate two *new* vectors of pairs with `.partition()`, then throw them all away when we move the pairs into the map.

Can we do this inline and avoid all these allocations?

Reply:
Fixed here: https://github.com/rust-bitcoin/rust-bitcoin/pull/103/commits/4efb72bba6a0846e0db5c2b2d3355bb0882dfce5#diff-6e12ff92fec0d0252713934f51fe80fcR104

Reply:
New construction [here](https://github.com/rust-bitcoin/rust-bitcoin/pull/103/commits/76a91e12502f6f837e9fa740233f049600f0e737#diff-6e12ff92fec0d0252713934f51fe80fcR104): only 1 allocation, needed because we need to collect all unknowns until we know if we have an unsigned transaction or not.